File: controllers/nao_team_1/FieldPlayer.java
//-----------------------------------------------------------------------------
//  File:         FieldPlayer.java (to be used in a Webots java controllers)
//  Date:         April 30, 2008
//  Description:  Field player "2", "3" or "4" for "red" or "blue" team
//  Project:      Robotstadium, the online robot soccer competition
//  Author:       Yvan Bourquin - www.cyberbotics.com
//  Changes:      November 4, 2008: Adapted to Webots6
//-----------------------------------------------------------------------------

import com.cyberbotics.webots.controller.*;

public class FieldPlayer extends Player {

  private Motion backwardsMotion, forwardsMotion, forwards50Motion, turnRight40Motion, turnLeft40Motion;
  private Motion turnRight60Motion, turnLeft60Motion, turnLeft180Motion, sideStepRightMotion, sideStepLeftMotion;

  private double goalDir = 0.0; // interpolated goal direction (with respect to front direction of robot body)

  public FieldPlayer(int playerID, int teamID) {
    super(playerID, teamID);
    backwardsMotion     = new Motion("../../motions/Backwards.motion");
    forwardsMotion      = new Motion("../../motions/Forwards.motion");
    forwards50Motion    = new Motion("../../motions/Forwards50.motion");
    turnRight40Motion   = new Motion("../../motions/TurnRight40.motion");
    turnLeft40Motion    = new Motion("../../motions/TurnLeft40.motion");
    turnRight60Motion   = new Motion("../../motions/TurnRight60.motion");
    turnLeft60Motion    = new Motion("../../motions/TurnLeft60.motion");
    turnLeft180Motion   = new Motion("../../motions/TurnLeft180.motion");
    sideStepRightMotion = new Motion("../../motions/SideStepRight.motion");
    sideStepLeftMotion  = new Motion("../../motions/SideStepLeft.motion");

    // move arms along the body
    Servo leftShoulderPitch = getServo("LShoulderPitch");
    leftShoulderPitch.setPosition(1.5);
    Servo rightShoulderPitch = getServo("RShoulderPitch");
    rightShoulderPitch.setPosition(1.5);
  }

  // normalize angle between -PI and +PI
  private double normalizeAngle(double angle) {
    while (angle > Math.PI) angle -= 2.0 * Math.PI;
    while (angle < -Math.PI) angle += 2.0 * Math.PI;
    return angle;
  }

  // relative body turn
  private void turnBodyRel(double angle) {
    if (angle > 0.7)
      turnRight60();
    else if (angle < -0.7)
      turnLeft60();
    else if (angle > 0.3)
      turnRight40();
    else if (angle < -0.3)
      turnLeft40();
  }

  protected void runStep() {
    super.runStep();
    double dir = camera.getGoalDirectionAngle();
    if (dir != SimpleCam.UNKNOWN)
      goalDir = dir - headYaw.getPosition();
  }

  private void turnRight60() {
    playMotion(turnRight60Motion); // 59.2 degrees
    goalDir = normalizeAngle(goalDir - 1.033);
  }

  private void turnLeft60() {
    playMotion(turnLeft60Motion); // 59.2 degrees
    goalDir = normalizeAngle(goalDir + 1.033);
  }

  private void turnRight40() {
    playMotion(turnRight40Motion); // 39.7 degrees
    goalDir = normalizeAngle(goalDir - 0.693);
  }

  private void turnLeft40() {
    playMotion(turnLeft40Motion); // 39.7 degrees
    goalDir = normalizeAngle(goalDir + 0.693);
  }
  
  private void turnLeft180() {
    playMotion(turnLeft180Motion); // 163.6 degrees
    goalDir = normalizeAngle(goalDir + 2.855);
  }

  @Override public void run() {
    step(SIMULATION_STEP);

    while (true) {

      runStep();

      getUpIfNecessary();

      while (getBallDirection() == SimpleCam.UNKNOWN) {
        //System.out.println("searching the ball"); 
        getUpIfNecessary();
        if (getBallDirection() != SimpleCam.UNKNOWN) break;
        headScan();
        if (getBallDirection() != SimpleCam.UNKNOWN) break;
        playMotion(backwardsMotion);
        if (getBallDirection() != SimpleCam.UNKNOWN) break;
        headScan();
        if (getBallDirection() != SimpleCam.UNKNOWN) break;
        turnLeft180();
      }

      double ballDir = getBallDirection();
      double ballDist = getBallDistance();

      //System.out.println("ball dist: " + ballDist + " ball dir: " + ballDir + " goal dir: " + goalDir);

      if (ballDist < 0.3) {
        //System.out.println("short distance");

        if (ballDir < -0.15)
          playMotion(sideStepLeftMotion);
        else if (ballDir > 0.15)
          playMotion(sideStepRightMotion);
        else if (goalDir < -0.35)
          turnLeft40();
        else if (goalDir > 0.35)
          turnRight40();
        else {
          //System.out.println("shooting !!!");
          playMotion(forwards50Motion);
        }
      }
      else {
        //System.out.println("long distance");
        double goDir = normalizeAngle(ballDir - goalDir);

        if (goDir < ballDir - 0.5)
          goDir = ballDir - 0.5;
        else if (goDir > ballDir + 0.5)
          goDir = ballDir + 0.5;

        goDir = normalizeAngle(goDir);

        turnBodyRel(goDir);
        if (ballDist < 0.6)
          playMotion(forwardsMotion);
        else
          playMotion(forwards50Motion);
      }
    }
  }
}

File: controllers/MyControl/MyControl/FieldPlayer.cpp
#include "FieldPlayer.hpp"
#include "NaoCam.hpp"
#include <webots/utils/Motion.hpp>
#include <webots/Servo.hpp>
#include <cmath>
#include <stdio.h>
#include <iostream>
using namespace webots;

FieldPlayer::FieldPlayer(int playerID, int teamID) : Player(playerID, teamID) {
  goalDir = 0.0; // interpolated goal direction (with respect to front direction of robot body)

  // move arms down
  getServo("LShoulderPitch")->setPosition(1.5);
  getServo("RShoulderPitch")->setPosition(1.5);
}

FieldPlayer::~FieldPlayer() {
  delete backwardsMotion;
  delete forwardsMotion;
  delete forwards50Motion;
  delete turnRight40Motion;
  delete turnLeft40Motion;
  delete turnRight60Motion;
  delete turnLeft60Motion;
  delete turnLeft180Motion;
  delete sideStepRightMotion;
  delete sideStepLeftMotion;
}

// normalize angle between -pi and +pi
static double normalizeAngle(double angle) {
  while (angle > M_PI) angle -= 2.0 * M_PI;
  while (angle < -M_PI) angle += 2.0 * M_PI;
  return angle;
}

// relative body turn
void FieldPlayer::turnBodyRel(double angle) {
  if (angle > 0.7)
    turnRight60();
  else if (angle < -0.7)
    turnLeft60();
  else if (angle > 0.3)
    turnRight40();
  else if (angle < -0.3)
    turnLeft40();
}

void FieldPlayer::runStep() {
  Player::runStep();
  double dir = camera->getGoalDirectionAngle();
  if (dir != NaoCam::UNKNOWN)
    goalDir = dir - headYaw->getPosition();
}

void FieldPlayer::turnRight60() {
  playMotion(turnRight60Motion); // 59.2 degrees
  goalDir = normalizeAngle(goalDir - 1.033);
}

void FieldPlayer::turnLeft60() {
  playMotion(turnLeft60Motion); // 59.2 degrees
  goalDir = normalizeAngle(goalDir + 1.033);
}

void FieldPlayer::turnRight40() {
  playMotion(turnRight40Motion); // 39.7 degrees
  goalDir = normalizeAngle(goalDir - 0.693);
}

void FieldPlayer::turnLeft40() {
  playMotion(turnLeft40Motion); // 39.7 degrees
  goalDir = normalizeAngle(goalDir + 0.693);
}

void FieldPlayer::turnLeft180() {
  playMotion(turnLeft180Motion); // 163.6 degrees
  goalDir = normalizeAngle(goalDir + 2.855);
}

void FieldPlayer::run() {
  step(SIMULATION_STEP);

  while (true) 
  {
    runStep();
    getUpIfNecessary();

    while (getBallDirection() == NaoCam::UNKNOWN)
	{
      std::cout << "searching the ball" << std::endl;
      getUpIfNecessary();
      if (getBallDirection() != NaoCam::UNKNOWN) break;
      headScan();
      if (getBallDirection() != NaoCam::UNKNOWN) break;
      playMotion(backwardsMotion);
      if (getBallDirection() != NaoCam::UNKNOWN) break;
      headScan();
      if (getBallDirection() != NaoCam::UNKNOWN) break;
      turnLeft180();
    }

    double ballDir = getBallDirection();
    double ballDist = getBallDistance();
	this->CalcPlayerState();

	//SendGPSPosition();
	//ReceiveGPSPosition();
	//CheckStage();
	doAction();


  }
}
void FieldPlayer::doAction()
{
	int current_action;
    double ballDist = getBallDistance();
	current_action=Kick/*world_info->pActiveScenario->ChooseActionForPlayer(this->playerID,this->nStage)*/;
	switch(current_action)
	{
	case Kick:
		//printf("ID #%i(st: #%i)ch kick [BD=%f]\n",this->playerID,this->nStage,ballDist);
		//fprintf(stdout,"ID #%i(st: #%i)ch kick [BD=%f]\n",this->playerID,this->nStage,ballDist);
		playMotion(shoot);
		break;
	case MoveForward:
		//printf("ID #%i(st: #%i) ch move f [BD=%f]\n",this->playerID,this->nStage,ballDist);
		playMotion(forwards50Motion);
		break;
	case Wait:
		//printf("ID #%i(st: #%i) ch wait [BD=%f]\n",this->playerID,this->nStage,ballDist);
		break;
	case DirectPass:
		playMotion(shoot);
		//printf("ID #%i(st: #%i) ch pass [BD=%f]\n",this->playerID,this->nStage,ballDist);
		break;
	case MoveSideLeft:
		playMotion(sideStepLeftMotion);
		//printf("ID #%i(st: #%i) ch move L [BD=%f]\n",this->playerID,this->nStage,ballDist);
		break;
	}
}

void FieldPlayer::CheckStage()
{
	double ballDist = getBallDistance();
	const double *Coord=this->gps->getValues();
	switch(playerID)
	{
	case 1:
		switch(nStage)
		{
		case 1:
			if(ballDist<0.2)
				nStage=2;
			break;
		case 2:
			if(ballDist>0.4)
				nStage=3;
			break;
		}
	break;
	case 2:
		switch(nStage)
		{
		case 1:
			if(ballDist<0.1)
				nStage=2;
			break;
		case 2:
			if(ballDist>0.4)
				nStage=3;
		}
		break;
	case 3:
		switch(nStage)
		{
		case 1:
			if(Coord[2]>=0)
			{
				nStage=2;
			}
			break;
		default:
			break;
		}
		break;
	}
}

void FieldPlayer::ReceiveGPSPosition()
{
	while (receiver->getQueueLength() > 0) 
	{
		const void *v= receiver->getData();
	
		char *t = new char[40];
		memcpy(t,v,40);
		/*printf("ID%i receive",playerID);*/
		//for(int i=0;i<40;i++)
		//	printf("%c",t[i]);
		//this->SaveInfoMessage(1,t);
	/*	printf("\n");*/
	}
}
void FieldPlayer::SendGPSPosition()
{
	const double *Coord=this->gps->getValues();
	char *t= new char[40];
	//sprintf(t,"ID%i\tX:%4f\tY:%4f\tZ:%4f\n",this->playerID,Coord[0],Coord[1],Coord[2]);
	//Message *nm = new Message(t,MessagePriority::medium,MessageType::mtGeoInform);
	//this->SaveInfoMessage(0,nm->getMessage());
	//emitter->send(t,40);
	
}

